enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/SlottedPipeBuilder.scala

```diff
@@ -355,12 +355,22 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
       case Union(_, _) =>
         val lhsInfo = pipelines(lhs.id)
         val rhsInfo = pipelines(rhs.id)
-
-        UnionSlottedPipe(lhs, rhs, lhsInfo, rhsInfo, pipeline, computeUnionSlots(lhsInfo, rhsInfo, pipeline))(id = id)
+        //val lhsOverlap = overlaps(pipeline, lhsInfo) // since pipeline is constructed from lhs the overlap will always be true
```

Comment:
This is not true, `MATCH (n) RETURN n UNION RETURN 42 AS n`
will (at least in the future) have a long slot for n in the lhs but a refslot in the outgoing pipeline

Reply:
We will need to have something like

```
val lhsOverlap = overlaps(pipeline, lhsInfo)
val rhsOverlap = overlaps(pipeline, rhsInfo)
if (lhsOverlap && rhsOverlap) FastUnionSlottedPipe(....)
else if (lhsOverlap) FastLeftUnionSlottedPipe(...)
else if (rhsOverlap) FastRightUnionSlottedPipe(...)
else UnionSlottedPipe(...)
```